### PR TITLE
Fix some issues edge-node cluster transition failing

### PR DESCRIPTION
**Pull request by naiming-zededa**
# Description

- Sometimes the node transitioning from single-node to cluster-mode
  failed due to can not verify the server CA. Keep restarting the k3s
  server will keep failing and requires a reboot
- Here we have a mechanism, not always reboot, only if the transitioning
  time has passed a limit and still can not join, then resort to reboot
- sometimes https://$join_serverIP:6443 fails does not tell if the ip
  level is even reachable or not. add the ping status also

## How to test and validate this PR

The issue is seen as a corner case
1. load the devices with kubevirt image
2. provision the edge-node cluster on the devices
3. make sure all the nodes eventually joins the cluster, this patch introduces a 10 minutes reboot to resolve a rare condition

## Changelog notes

Fix some issues edge-node cluster transition failing

## PR Backports

14.5: no, the feature is not officially part of that version
13.4: no, the feature is not officially part of that version

## Checklist

- [x] I've provided a proper description
- [ ] I've added the proper documentation
- [x] I've tested my PR on amd64 device
- [ ] I've tested my PR on arm64 device
- [x] I've written the test verification instructions
- [ ] I've set the proper labels to this PR

And the last but not least:

- [x] I've checked the boxes above, or I've provided a good reason why I didn't
  check them.

### CODE REVIEW

**Comment by OhmSpectator** [comment]
Do the same issues exist on 14.5-stable?

File: pkg/kube/cluster-init.sh [hunk 1]
```diff
@@ -720,6 +720,11 @@ EOF
         # Check if the join Server is available by kubernetes, wait here until it is ready
         counter=0
         touch "$CLUSTER_WAIT_FILE"
+
+        # Initialize ping counters before the loop (add this near the top of your function/script)
```
Comment (OhmSpectator) [inline pkg/kube/cluster-init.sh]: You have already added it, you can remove this part of the comment =)

File: pkg/kube/cluster-init.sh [hunk 2]
```diff
@@ -596,6 +597,18 @@ check_cluster_config_change() {
...
             logmsg "provision config file for node to cluster mode"
             provision_cluster_config_file true
+            if [ "$is_bootstrap" = "false" ]; then
+              # we got here because we know the bootstrap node is already running
+              # For a non-bootstrap node, create transition file and record timestamp
+              # This will be checked by check_cluster_transition_done function
+              # We have seen in some cases, the k3s server on this node can not join the cluster
+              # due to CA cert issues, and reboot is needed to get out of this state
+              # so we do not always reboot here, only if it is needed
+              date +%s > /var/lib/transition-to-cluster
```
Comment (OhmSpectator) [inline pkg/kube/cluster-init.sh]: I would like to understand how this works. From which container are we running this code? How is `/var/lib` mounted in this context? Is it part of the host filesystem or is it a temporary filesystem (tmpfs)? What is the lifespan of the directory and files in this case? Do they persist between node reboots?

File: pkg/kube/cluster-init.sh [hunk 3]
```diff
@@ -622,10 +635,53 @@ check_cluster_config_change() {
...
+    transition_timestamp=$(cat /var/lib/transition-to-cluster)
+    current_timestamp=$(date +%s)
+    elapsed_time=$((current_timestamp - transition_timestamp))
+
+    if [ "$elapsed_time" -ge 600 ]; then # 10 minutes in seconds
+        logmsg "Cluster transition timeout: Been waiting for ${elapsed_time} seconds"
+        # Update timestamp and reboot
+        date +%s > /var/lib/transition-to-cluster
+        logmsg "Rebooting system to retry cluster transition..."
+        reboot
```
Comment (OhmSpectator) [inline pkg/kube/cluster-init.sh]: Am I correct in saying that if it fails repeatedly, we will reboot every 10 minutes? Should we limit the number of attempts?

File: pkg/kube/cluster-init.sh (hunk 1 above)
Comment (naiming-zededa) [reply pkg/kube/cluster-init.sh]: sure.

File: pkg/kube/cluster-init.sh (hunk 2 above)
Comment (naiming-zededa) [reply pkg/kube/cluster-init.sh]: this mounts it to persist, in the 'kube' container. it persists across reboots.
```
mount_etcd_vol() {
        # NOTE: We only support zfs storage in production systems because data is persisted on zvol.
        # This is formatted in vaultmgr
        logmsg "Wait for persist/etcd-storage zvol"
        while [ ! -b /dev/zvol/persist/etcd-storage ];
        do
                sleep 1
        done
        mount /dev/zvol/persist/etcd-storage /var/lib  ## This is where we persist the cluster components (etcd)
        logmsg "persist/etcd-storage available"
}
```

Comment (OhmSpectator) [reply pkg/kube/cluster-init.sh]: Thanks for the clarification!

File: pkg/kube/cluster-init.sh (hunk 3 above)
Comment (naiming-zededa) [reply pkg/kube/cluster-init.sh]: I updated to limit reboots at 3 times max.

**Comment by OhmSpectator** [comment]
> I'm not sure. We have a number of fixes in this batch after QA testings.

Please check the code in 14.5-stable then. We need to know what to backport.

**Comment by OhmSpectator** [comment]
> > I'm not sure. We have a number of fixes in this batch after QA testings.
> 
> Please check the code in 14.5-stable then. We need to know what to backport.

@naiming-zededa, did you check the code in 14.5?

**Comment by naiming-zededa** [comment]
> > > I'm not sure. We have a number of fixes in this batch after QA testings.
> > 
> > 
> > Please check the code in 14.5-stable then. We need to know what to backport.
> 
> @naiming-zededa, did you check the code in 14.5?

the fix will still apply, but not direct cherry-pick.
